### ISSUE TITLE
Inconsistent (and intransparent) handling of groups containing duplicates

### ISSUE THREAD

**Issue by zemanj** [MDAnalysis/mdanalysis#1911]
### Expected behaviour

Calling `center_of_mass(pbc=False)` on a group containing duplicates should behave similar to calling `center_of_mass(pbc=True)` on the same group. The only difference should be whether PBC are taken into account. Furthermore, the way duplicates are handled should be transparent to the user.

### Actual behaviour
Calling `center_of_mass(pbc=False)` on a group containing duplicates works. However, it isn't transparent to the user if and how duplicates are handled.

Calling `center_of_mass(pbc=True)` on a group containing duplicates raises a `ValueError`:
```python
ValueError: Length of weights not compatible with specified axis.
```

### Code to reproduce the behaviour

``` python
import MDAnalysis as mda
from MDAnalysisTests.datafiles import PSF, DCD

# Load universe and assign a valid box:
u = mda.Universe(PSF, DCD)
u.dimensions[:] = [50, 50, 50, 90, 90, 90]

# Make a selection with duplicates:
selection = u.select_atoms("resname ALA")
selection += selection.residues[:selection.n_residues//2].atoms

# Compute COM without PBC:
selection.center_of_mass(pbc=False)  # This runs without complaints

# Compute COM with PBC taken into account:
selection.center_of_mass(pbc=True)  # Raises ValueError
```

### Code section causing the behavior:
`center_of_mass(pbc=True)` calls `center()`, which, in turn, calls `pack_into_box()`, where coordinates of *only unique atoms* are returned. The line causing trouble is [this one](https://github.com/MDAnalysis/mdanalysis/blob/5feb9f4ffc7434f18eed85d517912c1cde4a1ac7/package/MDAnalysis/core/groups.py#L945).

### Current version of MDAnalysis:
0.18.1-dev

### Notes
- Affects PR #1903
- Other parts of the code such as `center_of_geometry()` etc. are probably affected as well.

**Comment by richardjgowers** [comment]
Yeah I'm not really sure what the expected behaviour is for this.  I guess if you've got duplicates of an atom it should weight that atom more, but I can't really think of a sane use for that...  Definitely worth raising a warning about either way

**Comment by zemanj** [comment]
I just checked which classes use `pack_into_box()`. Luckily, there are only: `core.groups.GroupBase` and `core.topologyattrs.Masses`.

Methods of `core.groups.GroupBase` using `pack_into_box()` (together with my two cents):
- `center()`: inconsistently affected by duplicate atoms, should be consistent
- `bbox()`: not affected by duplicates
- `bsphere()`: inconsistently affected by duplicate atoms, but should never be affected
- `wrap()`: inconsistently affected by duplicate atoms, but should be consistent

And methods of `core.topologyattrs.Masses`:
- `moment_of_inertia()`: inconsistently affected by duplicate atoms, should be consistent
- `radius_of_gyration()`: inconsistently affected by duplicate atoms, should be consistent
- `shape_parameter()`: inconsistently affected by duplicate atoms, should be consistent
- `asphericity()`: inconsistently affected by duplicate atoms, should be consistent

Proposed solution:
- Always perform calculations on `group.<member>.unique` (where `<member>` is usually to be replaced with `atoms`).
- Check if the group contains duplicates by comparing `group.atoms.n_atoms` with `group.atoms.unique.n_atoms` (unless there's a faster way). If there are duplicates, raise a `UserWarning` stating that only unique members are taken into account.

**Comment by richardjgowers** [comment]
@zemanj that all sounds sane.  It will be worth checking how long `.unique` takes in comparison to eg `center_of_mass` on a relatively large AG

**Comment by zemanj** [comment]
Regarding `core.groups.GroupBase.pack_into_box()` itself:
It makes perfect sense that the actual packing is only performed on unique atoms, anything else would be a waste of computing time. However, in my opinion, the returned positions should always have the same length as the calling group. Accomplishing both should be rather trivial.

**Comment by zemanj** [comment]
@richardjgowers That's a good point. At this point, `.unique` is called whenever PBC are to be taken into account. One could, however, disable this sort of duplicate checking on request, say, by yet another kwarg. ;-) Only if it potentially affects performance, of course. I'll probably have time to look into this tomorrow.

**Comment by zemanj** [comment]
@richardjgowers The `self.atoms.unique` part of `pack_into_box(inplace=True)` takes 20% of the runtime. For `inplace=False` (used more frequently in the code base), it's **32%**. See [here](https://gist.github.com/zemanj/9c50bd0fadd636ba4f75ce4ae94791c2) for details
That's pretty bad regarding the fact that AtomGroups are usually unique. (I assume that since nobody seems to have complained about a broken `pack_into_box()` method so far.)

**Comment by orbeckst** [comment]
@zemanj thank you for comprehensively looking into this. The timings https://github.com/MDAnalysis/mdanalysis/issues/1911#issuecomment-391384081 are frightening.

I can't think of good use cases where one would want to use duplication of atoms as weights – I think we often provide a `weights` kwarg when this makes sense. However, your point about returning the same shape https://github.com/MDAnalysis/mdanalysis/issues/1911#issuecomment-391134233 is important for consistency; it might happen rarely so perhaps we can have a standard code path that assumes uniqueness but we should not be overly prescriptive in how users can use groups.

**Comment by zemanj** [comment]
@orbeckst The timings only seem frightening to you because you haven't seen the same timings on `ResidueGroups` yet... :cry: But I'll save you from that until I come up with a proper solution, which I am working on right now.

**Comment by zemanj** [comment]
@MDAnalysis/coredevs
- Is there a specific reason for `AtomGroup.atoms` always returning a *new* `AtomGroup` instance instead of `self`?
- Likewise, do `ResidueGroup.atoms` and `SegmentGroup.atoms` always have to return a new instance? If not, one could cache and reuse the returned AtomGroup. This would have the potential to massively speed up repeated operations involving `*Group.atoms`.
- Furthermore, does `*Group.unique` have to return a new instance if the group is known to be unique?

I am asking all this because I'm currently looking into how to speed up methods that require *unique* AtomGroups. I already managed to cache the uniqueness of Groups, which is hard to preserve across implicit `GroupBase.__init__()` calls (and generally shouldn't be). If the `AtomGroup` returned by `*Group.atoms` were cached, that would also apply to its uniqueness property.

**Comment by jbarnoud** [comment]
Just my 2 cents about point 3: I remember that *Groups are mostly immutable, but are they completely so? If not, then we want to be sure the `unique` method behave the same regardless of the content of the group.

**Comment by zemanj** [comment]
@jbarnoud I think I figured that out (please correct me if I'm wrong!):
`GroupBase` inherits from `_MutableBase`, and the _MutableBase.__new__() constructor reuses class instances. However, the `GroupBase.__init__()` initializer is called whenever a group is changed. Consequently, that also applies to concatenation, slicing, updating selections, etc. Therefore, I currently have the cached properties being reset by that initializer, which seems to work fine.
The initializer is **not** called, however, when properties on a different topology level change, which complicates point 2, as @richardjgowers already mentioned.

**Comment by zemanj** [comment]
Since `AtomGroup`s are unique in most cases, it is generally faster to have a quick check for uniqueness before attempting to actually create a unique `AtomGroup`. I tested a couple of methods to accomplish just that. This is what I came up with:
1. The most obvious way which actually *does* create a unique `AtomGroup`:
```Python
def is_unique_mda(ag):
    return len(ag) == len(ag.unique)
```
2. Using `numpy.unique()` on `AtomGroup.ix` (avoids creating a new AG instance):
```Python
def is_unique_numpy(ag):
    ix = ag.ix
    u_ix = np.unique(ix)
    return len(u_ix) == len(ix)
```
3. Checking adjacent elements of a sorted copy of `AtomGroup.ix` for inequality and returning the minimum of the resulting boolean array:
```Python
def is_unique_min(ag):
    if len(ag) <= 1:
        return True
    s_ix = np.sort(ag.ix)
    return (s_ix[1:] != s_ix[:-1]).min()
```
4. Same as above, but use numpy's lazily (only for `np.bool_` afaik) implemented `ndarray.argmin()`:
```Python
def is_unique_argmin(ag):
    if len(ag) <= 1:
        return True
    s_ix = np.sort(ag.ix)
    mask = s_ix[1:] != s_ix[:-1]
    return mask[mask.argmin()]
```

See [here](https://gist.github.com/zemanj/60c04957385322ec347a751506b6f4c5) for timings.
Long story short: Method 4 (`is_unique_argmin()`) is fastest.

**Comment by zemanj** [comment]
Further methods affected by duplicates are:
- `total_mass()`
- `total_charge()`

**Comment by orbeckst** [comment]
@zemanj does PR #1922 fix `total_mass()` etc as mentioned in https://github.com/MDAnalysis/mdanalysis/issues/1911#issuecomment-395841576 as well?

**Comment by orbeckst** [comment]
PR #1911 said in the description that it fixes this issue so I am closing this one. Please re-open if this is incorrect.
